server/src/main/java/org/opensearch/cluster/routing/remote/RemoteRoutingTableService.java

```diff
@@ -0,0 +1,243 @@
+    private final Supplier<RepositoriesService> repositoriesService;
+    private BlobStoreRepository blobStoreRepository;
+
+    public RemoteRoutingTableService(Supplier<RepositoriesService> repositoriesService, Settings settings) {
+        assert isRemoteRoutingTableEnabled(settings) : "Remote routing table is not enabled";
+        this.repositoriesService = repositoriesService;
+        this.settings = settings;
+    }
+
+    public List<IndexRoutingTable> getChangedIndicesRouting(ClusterState previousClusterState,
+                                                            ClusterState clusterState) {
+        Map<String, IndexRoutingTable> previousIndexRoutingTable = previousClusterState.getRoutingTable().getIndicesRouting();
+        List<IndexRoutingTable> changedIndicesRouting = new ArrayList<>();
+        for (IndexRoutingTable indexRouting : clusterState.getRoutingTable().getIndicesRouting().values()) {
+            if (!(previousIndexRoutingTable.containsKey(indexRouting.getIndex().getName()) && indexRouting.equals(previousIndexRoutingTable.get(indexRouting.getIndex().getName())))) {
+                changedIndicesRouting.add(indexRouting);
+                logger.info("changedIndicesRouting {}", indexRouting.prettyPrint());
+            }
+        }
+
+        return changedIndicesRouting;
+    }
+
+    public CheckedRunnable<IOException> getIndexRoutingAsyncAction(
+        ClusterState clusterState,
+        IndexRoutingTable indexRouting,
+        LatchedActionListener<ClusterMetadataManifest.UploadedMetadata> latchedActionListener,
+        BlobPath clusterBasePath
+    ) throws IOException {
+
+        BlobPath indexRoutingPath = clusterBasePath.add(INDEX_ROUTING_PATH_TOKEN);
+        logger.info("custerMetadataBasePath {}", indexRoutingPath);
+
+        BlobPath path = RemoteStoreEnums.PathType.HASHED_PREFIX.path(RemoteStorePathStrategy.BasePathInput.builder()
+                .basePath(indexRoutingPath)
+                .indexUUID(indexRouting.getIndex().getUUID())
+                .build(),
+            RemoteStoreEnums.PathHashAlgorithm.FNV_1A_BASE64);
+        logger.info("path from prefix hasd  {}", path);
+        final BlobContainer blobContainer = blobStoreRepository.blobStore().blobContainer(path);
+
+        final String fileName = getIndexRoutingFileName();
+        logger.info("fileName {}", fileName);
+
+        ActionListener<Void> completionListener = ActionListener.wrap(
+            resp -> latchedActionListener.onResponse(
+                new ClusterMetadataManifest.UploadedIndexMetadata(
+                    indexRouting.getIndex().getName(),
+                    indexRouting.getIndex().getUUID(),
+                    path.buildAsString() + fileName,
+                    INDEX_ROUTING_METADATA_PREFIX
+                )
+            ),
+            ex -> latchedActionListener.onFailure(new RemoteClusterStateService.RemoteStateTransferException(indexRouting.getIndex().toString(), ex))
```

Comment:
lets add a error msg instead of just index name to provide better visibility while debugging

Reply:
Done